Logging overhaul

obackhouse opened the pull request:
This PR overhauls the logging system:

- `Base` no longer inherits from `pyscf.lib.StreamObject`, so we don't need to use their standards for the logging internals
- Added `rich` dependency
- Changed logging internals to use a `rich.Console` instance with colour coded outputs, live status and table updates, and much better formatted outputs
- Added `MOMENTGW_SILENT` environment variable (or `momentGW.logging.silent`) to suppress outputs
- Removed timings from outputs -- a table of useful timings can be printed using `momentGW.logging.dump_times()` at the end of a script
- Removed `mo_energy` and `mo_coeff` from kernel arguments -- this is legacy stuff from PySCF styles, I don't think it's necessary and doesn't even work properly

To-do:

- [x] `momentGW`
- [x] `momentGW.uhf`
- [x] `momentGW.pbc`
- [x] `momentGW.pbc.uhf`
- [x] Clean up

obackhouse commented on momentGW/base.py:
```diff
@@ -3,52 +3,90 @@
+        init_logging()
+        logging.write("")
+        logging.write(f"[bold underline]{self.name}[/]", comment="Initialisation of solver")
+        table = logging.Table(title="Options")
```
move options logging inside `kernel`

mkakcl commented:
Potential logging changes
- have N/A if no OMP_NUM_Thread - Not sure how hard it is do but running it locally has no output
- conv_logical output is <built-in function all> - not sure what it should be but obviously not super useful information
- Have it display PBC somewhere - like can intuit that it is a PBC calc from gamma's in places, but probably good to have somewhere
- k-point mesh size - realise this is more from pyscf and is probably in their output somewhere, but having the k-mesh size would be very useful in mGW.

mkakcl commented:
Ideas/comment on the logger
- Install rich as part of pip -  I seem to recall you mentioning you would deploy to pip again, so probably not worth making a comment of it in the read me, but just for the future.
- Another line before momentGW header may be nice - Pyscf is so messy, just nice to have a clearer split
- Versions/git hashes for scipy and H5py? - less essential to methods but only other real packages we use
- Would be nice to have non-default values in bold/other colour to see what you have inputted vs default
- Chemical potential output has no colour, expected/desired? Just since most values out are a colour.
- Would an iteration number or a measure of iterations to convergence be useful? - like "Convergence after 16/50 iterations" where 50 is the max you allow

obackhouse commented:
> * [x]  have N/A if no OMP_NUM_Thread - Not sure how hard it is do but running it locally has no output

Changed it to report `OpenMP threads:` rather than explicitly the environment variable

> * [ ]  conv_logical output is  - not sure what it should be but obviously not super useful information

Changed `callable` and `type` formatting to use `__name__` instead of `repr`

> * [ ]  Have it display PBC somewhere - like can intuit that it is a PBC calc from gamma's in places, but probably good to have somewhere

Happy as-is with the below change, the solver name also has `k` in it

> * [x]  k-point mesh size - realise this is more from pyscf and is probably in their output somewhere, but having the k-mesh size would be very useful in mGW.

Added a size table in the header with the orbitals and k-points

obackhouse commented:
> * [ ]  Another line before momentGW header may be nice - Pyscf is so messy, just nice to have a clearer split

I think i'll avoid this one, because then there will be an awkward blank line if you run with `verbose=0`. Combining logging of different packages should be a script-side thing, IMO

> * [x]  Versions/git hashes for scipy and H5py? - less essential to methods but only other real packages we use

Done

> * [ ]  Would be nice to have non-default values in bold/other colour to see what you have inputted vs default

I've made default options dim and modified ones non-dim. I'm not sure i like this, because they're almost always going to be default so it just detracts from the output... I tried adding a "Modified" column and didn't like that much either. Thoughts @mkakcl ?

> * [ ]  Chemical potential output has no colour, expected/desired? Just since most values out are a colour.

It's not an output and has no good/bad status (only consequences that can be good/bad), so it's not coloured. My colour scheme is:
 - green: good
 - yellow: ok
 - red: bad
 - cyan: input
 - blue: output energy (ground state or excitation)

> * [ ]  Would an iteration number or a measure of iterations to convergence be useful? - like "Convergence after 16/50 iterations" where 50 is the max you allow

i'll pass on this unless you have strong feelings -- it means storing an extra state in the object about how many iterations passed. I don't think that's necessarily a bad thing if it's done cleanly but we don't have anything to track the history of the convergence post-hoc

mkakcl commented:
On the output changes, I personally quite like it when there is a difference, particularly when having default values dim and change brighter. To me, it just makes it clearer what you have changed. Get the point about the flow/overall look changes. But to me, it just brings your eye to it, in the same way the colours do, which is nice when looking through an output and figuring out what I did.

obackhouse commented:
yeah i like it now, i just made the option name dim rather than the coloured value as well

mkakcl reviewed (approved):
Changes look great, nice increase to the clarity of the output from the logger. Merge when ready.
